contracts/Market.sol

```diff
@@ -1929,6 +1771,168 @@ contract Market is
     return (countOfRemovalsAllocated, ids, amounts);
   }
 
+  /**
+   * @notice Allocates the removals, amounts, and suppliers needed to fulfill the purchase drawing only from
+   * the vintages specified.
+   * @param amount The number of carbon removals to purchase.
+   * @param vintages The vintages from which to purchase carbon removals.
+   * @return countOfRemovalsAllocated The number of distinct removal IDs used to fulfill this order.
+   * @return ids An array of the removal IDs being drawn from to fulfill this order.
+   * @return amounts An array of amounts being allocated from each corresponding removal token.
+   * @return suppliers The address of the supplier who owns each corresponding removal token.
+   */
+  function _allocateSupplySpecificVintages(
+    uint256 amount,
+    uint256[] memory vintages
+  )
+    private
+    returns (
+      uint256 countOfRemovalsAllocated,
+      uint256[] memory ids,
+      uint256[] memory amounts,
+      address[] memory suppliers
+    )
+  {
+    uint256 remainingAmountToFill = amount;
+    uint256 countOfListedRemovals = _removal.numberOfTokensOwnedByAddress({
+      account: address(this)
+    });
+    ids = new uint256[](countOfListedRemovals);
+    amounts = new uint256[](countOfListedRemovals);
+    suppliers = new address[](countOfListedRemovals);
+    countOfRemovalsAllocated = 0;
+    for (uint256 i = 0; i < countOfListedRemovals; ++i) {
+      uint256 removalId = _listedSupply[_currentSupplierAddress]
+        .getNextRemovalForSaleFromVintages({vintages: vintages});
+      // if removalId is 0 here, this supplier doesn't have any removals from the vintages requested
+      // at that point we should simply move on to the next supplier
```

Comment:
I've explored this heavily in thinking about this algorithm, and I don't believe there is a better approach here for much the same reason that we have to loop in our regular fulfillment with the upper bound on iterations being the count of all available removal tokens for sale. our data structure is indexed first by supplier, then by vintage. there is overhead we could perform to figure out which suppliers have any of the valid vintages and how much they have which would be a tighter bound on our loop iterations, but it would require doing work during the overhead that we can actually potentially avoid by being lazier about it.  i.e. if we were to try to be smarter about this fulfillment, we still have to walk through our suppliers and examine their data structures to understand who has how much of which vintages we care about. but if we're lazier about it, we can basically start the same walk through the data structure with a high likelihood of NOT having to do the full walk that we would do if we were going to do some organizational overhead before performing fulfillment.  Here, we basically say, "maximum case is that all removal sin our marketplace meet the criteria and could be part of this fulfillment, so that's how we'll bound our fulfillment loop, but we'll look as efficiently as possible at each supplier and if they don't have any removal sfrom the vintages we care about, we'll move right on".  That's my best explanation.  If you have a specific idea for improving this I'm open it but I've thought about it a lot